### PR TITLE
Implement ContainExactly and MatchArray cops with autocorrect

### PR DESCRIPTION
**Pull request by faucct**
Here are some examples:

```ruby
# bad
it { is_expected.to contain_exactly(*array1, *array2) }

# good
it { is_expected.to match_array(array1 + array2) }

# bad
it { is_expected.to match_array([content1, content2]) }

# good
it { is_expected.to contain_exactly(content1, content2) }

# good
it { is_expected.to contain_exactly(content, *array) }

# good
it { is_expected.to match_array([content] + array) }
```

### CODE REVIEW

**Comment by backus** [comment]
Wow. I think I just never reviewed this. Sorry about that! I'll try to review soon

File: lib/rubocop/cop/rspec/contain_exactly.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      class ContainExactly < Cop
+        MSG = 'Prefer `match_array` when matching array values'.freeze
+
+        def on_send(node)
+          _receiver, method_name, *args = *node
+          return unless method_name == :contain_exactly
+
+          args.all? do |child_node|
+            child_node.is_a?(Node) && child_node.type == :splat
+          end && add_offense(node, :expression, MSG)
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/contain_exactly.rb]: the `&& add_offence` part looks weird. IMO it will make for a better style to have
```
args.any? do |child_node|
  return unless child_node.is_a?(Node) && child_node.type == :splat
end
```
then you can just add_offence at the end of the method

File: lib/rubocop/cop/rspec/match_array.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      class MatchArray < Cop
+        MSG = 'Prefer `contain_exactly` when matching an array literal'.freeze
+
+        def on_send(node)
+          _receiver, method_name, *args = *node
+          return unless method_name == :match_array
+
+          args.all? do |child_node|
+            child_node.is_a?(Node) && child_node.type == :array
+          end && add_offense(node, :expression, MSG)
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/match_array.rb]: same as for the other cop

File: spec/rubocop/cop/rspec/contain_exactly_spec.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+describe RuboCop::Cop::RSpec::ContainExactly do
+  subject(:cop) { described_class.new }
+
+  it 'finds `match_array` instead of an instance double' do
```
Comment (Darhazer) [inline spec/rubocop/cop/rspec/contain_exactly_spec.rb]: `instead of contain_exactly`. But I would reword it as 'flags `match_array` with only splat arguments'

File: lib/rubocop/cop/rspec/match_array.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      #   # bad
+      #   it { is_expected.to match_array([content1, content2]) }
+      #
+      #   # good
+      #   it { is_expected.to contain_exactly(content1, content2) }
+      #
+      #   # good
+      #   it { is_expected.to match_array([content] + array) }
+      class MatchArray < Cop
+        MSG = 'Prefer `contain_exactly` when matching an array literal'.freeze
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/match_array.rb]: Please add a dot at the end of the message

File: lib/rubocop/cop/rspec/contain_exactly.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      #   # bad
+      #   it { is_expected.to contain_exactly(*array1, *array2) }
+      #
+      #   # good
+      #   it { is_expected.to match_array(array1 + array2) }
+      #
+      #   # good
+      #   it { is_expected.to contain_exactly(content, *array) }
+      class ContainExactly < Cop
+        MSG = 'Prefer `match_array` when matching array values'.freeze
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/contain_exactly.rb]: Please add dot at the end of the message

File: lib/rubocop/cop/rspec/contain_exactly.rb (hunk 1 above)
Comment (Darhazer) [inline lib/rubocop/cop/rspec/contain_exactly.rb]: You can skip the MSG argument of add_offense

File: spec/rubocop/cop/rspec/contain_exactly_spec.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    expect_violation(<<-RUBY)
+      it { is_expected.to contain_exactly(*array1, *array2) }
+                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Prefer `match_array` when matching array values
+    RUBY
+  end
+
+  it 'finds `match_array` instead of an instance double' do
```
Comment (Darhazer) [inline spec/rubocop/cop/rspec/contain_exactly_spec.rb]: please provide proper description

File: spec/rubocop/cop/rspec/contain_exactly_spec.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  it 'finds `expect_any_instance_of` instead of an instance double' do
+    expect_no_violations(<<-RUBY)
+      it { is_expected.to match_array([content] + array) }
+    RUBY
+  end
+
+  it 'finds `expect_any_instance_of` instead of an instance double' do
+    expect_no_violations(<<-RUBY)
+      it { is_expected.to contain_exactly(content, *array) }
```
Comment (Darhazer) [inline spec/rubocop/cop/rspec/contain_exactly_spec.rb]: to match_array

File: spec/rubocop/cop/rspec/match_array_spec.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  it 'finds `match_array` instead of an instance double' do
+    expect_violation(<<-RUBY)
+      it { is_expected.to match_array([content1, content2]) }
+                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Prefer `contain_exactly` when matching an array literal
+    RUBY
+  end
+
+  it 'finds `match_array` instead of an instance double' do
+    expect_no_violations(<<-RUBY)
+      it { is_expected.to contain_exactly(content1, content2) }
```
Comment (Darhazer) [inline spec/rubocop/cop/rspec/match_array_spec.rb]: to match_array

File: spec/rubocop/cop/rspec/match_array_spec.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  it 'finds `expect_any_instance_of` instead of an instance double' do
+    expect_no_violations(<<-RUBY)
+      it { is_expected.to match_array([content] + array) }
+    RUBY
+  end
+
+  it 'finds `expect_any_instance_of` instead of an instance double' do
+    expect_no_violations(<<-RUBY)
+      it { is_expected.to contain_exactly(content, *array) }
```
Comment (Darhazer) [inline spec/rubocop/cop/rspec/match_array_spec.rb]: to match_array

**Comment by dgollahon** [comment]
@faucct: Are you interested in updating / finishing this PR? Or would you like me or someone else to take it over?

**Comment by faucct** [comment]
Rebased and respected the comments.

File: lib/rubocop/cop/rspec/contain_exactly.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+      #   # good
+      #   it { is_expected.to match_array(array1 + array2) }
+      #
+      #   # good
+      #   it { is_expected.to contain_exactly(content, *array) }
+      class ContainExactly < Cop
+        MSG = 'Prefer `match_array` when matching array values.'.freeze
+
+        def on_send(node)
+          _receiver, method_name, *args = *node
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/contain_exactly.rb]: a bit cleaner implementation is:
```
  return unless node.method_name == :contain_exactly
  return unless node.each_child_node.all?(&:splat_type?)

  add_offense node, :expression
```

File: lib/rubocop/cop/rspec/match_array.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+      #   # good
+      #   it { is_expected.to contain_exactly(content1, content2) }
+      #
+      #   # good
+      #   it { is_expected.to match_array([content] + array) }
+      class MatchArray < Cop
+        MSG = 'Prefer `contain_exactly` when matching an array literal.'.freeze
+
+        def on_send(node)
+          _receiver, method_name, *args = *node
```
Comment (Darhazer) [inline lib/rubocop/cop/rspec/match_array.rb]: a bit cleaner implementation
```
  return unless node.method_name == :match_array
  return unless node.each_child_node.all?(&:array_type?)

  add_offense node, :expression
```

**Comment by Darhazer** [comment]
@faucct can you add a changelog entry?

**Comment by Darhazer** [comment]
I have applied the needed changes in #464
